### PR TITLE
Event consumer for submission config reloading when a collection changes

### LINKED ISSUE #8633: New atribute in Item submission for name-mapping based on collection's entity type
**Is your feature request related to a problem? Please describe.**
For associating a specific submission form to an entity type, you are required to know the handle of the collection. That is how it worked in the past and currently works. But, if you load entities model, collections currently have a way to specify which Entity types it's related to ease and better suggest external services to use. But with item submission, that work isn't considered.

**Describe the solution you'd like**
We currently specify a submission form using the handle
```
<item-submission>
    <submission-map>
		<name-map collection-handle="12345/6789" submission-name="Person"/>

```
but I think this configuration can be extended to support a new parameter, like `collection-entity-type`:

```
<item-submission>
    <submission-map>
		<name-map collection-entity-type="Person" submission-name="Person"/>

```
An that will enable all configured collections with "Person"
image

to be associated with this `submission-name="Person"`.

**Describe alternatives or workarounds you've considered**
Not the best solution, but the most easiest to implement this is to change `getCollectionsBySubmissionConfig`, of `SubmissionConfigReader `class, and extend it in order to look for all collections that have that `entity-type` attribute and add those collections to the resulted List.

```java
    public List<Collection> getCollectionsBySubmissionConfig(Context context, String submitName)
        throws IllegalStateException, SQLException {
        List<Collection> results = new ArrayList<>();
        // get the submission-map keys
        for (String handle : collectionToSubmissionConfig.keySet()) {
            if (!DEFAULT_COLLECTION.equals(handle)) {
                if (collectionToSubmissionConfig.get(handle).equals(submitName)) {
                    DSpaceObject result = HandleServiceFactory.getInstance().getHandleService()
                                                              .resolveToObject(context, handle);
                    if (result != null) {
                        results.add((Collection) result);
                    }
                }
            }
        }
        return results;
    }
```

Also, processMap method should be changed in order to support `collection-entity-type` parameter. And item-submission.dtd should also accommodate this.

**Additional context**
There is an issue regarding the change of item-submission to Spring: https://github.com/DSpace/DSpace/issues/7985
There will be an issue when you have the same collection handle associated with different submission forms. I would suggest to give priority to the configuration that specified the handle and the that occurs first, discarding the rest.

### PR DESCRIPTION
**Pull request by paulo-graca**
## References
_Add references/links to any related issues or PRs. These may include:_
* fixes #8633
* complements #8855 (adds a way to reload configurations)
* partially addresses #7985 (it needs a refactor)

## Description
This PR is a complement to #8855. Enabling the system to have one single SubmissionConfigReader that can be reloaded when needed. To do it so, PR adds an Event Consumer that will run on configured actions for collections. 

## Instructions for Reviewers
This PR is easy to test. In order to do it so you will have to:
- Ensure you have this (or some other **collection-entity-type**) enabled on your dspace/config/item-submission.xml file
``` xml
        <name-map collection-entity-type="Publication" submission-name="Publication"/>
        <name-map collection-entity-type="Person" submission-name="Person"/>
        <name-map collection-entity-type="Project" submission-name="Project"/>
        <name-map collection-entity-type="OrgUnit" submission-name="OrgUnit"/>
```
- Also ensure you have `event.consumer.submissionconfig.filters` configured on your dspace.cfg
- Start/Restart your tomcat to assume this changes
- On any existing on new community, add a new collection and specify it's entity type (like: OrgUnit)
- Then create a new item on that collection

The expected result should be displayed a specific form for that entity type (like OrgUnit).

List of changes in this PR:
* This PR does some changes in the way to deal with SubmissionConfigReader
* It adds service and factory services for submissionConfigs for SubmissionConfigReader
* It adds a new event consumer submissionConfig that, in case of collection changes, it will reload item-submission configuration
* To deal with the fact that this PR - https://github.com/DSpace/DSpace/pull/8855 - uses Discovery to obtain the collection's data, and at the moment of collection creation, the indexing didn't occur yet, reloading the form before indexing don't consider the entity type. So, I force a Collection index action on the event consumer.

### CODE REVIEW

**Comment by blancoj** [comment]
@paulo-graca I'm wondering how complete this PR is?  I've made some customizations to DCInputsReader.java to create a pick list in the submission form based on the user logged in.  With the code I have now (7.6), the form always looks the same, no matter who is logged in.  I think if I put in this patch, the deposit form will be recreated when the user starts the submission.  I just don't know how complete this patch is.  Also, maybe I don't need the entire patch.  Are there particular lines that can be commented out or added that will force the form to be recreated when the user starts a submission?

File: dspace-api/src/main/java/org/dspace/submit/consumer/SubmissionConfigConsumer.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        if ( st == Constants.COLLECTION ) {
+            switch (et) {
+                case Event.MODIFY_METADATA:
+                    // Submission configuration it's based on solr
+                    // for collection's entity type but, at this point
+                    // that info isn't indexed yet, we need to force it
+                    DSpaceObject subject = event.getSubject(ctx);
+                    Collection collectionFromDSOSubject = (Collection) subject;
+                    indexer.indexContent(ctx, new IndexableCollection (collectionFromDSOSubject), true, false, false);
+                    indexer.commit();
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/submit/consumer/SubmissionConfigConsumer.java:57]: I'm a little confused by the need to force reindexing of the Collection into Solr.  I had always thought that the Event Consumers listed in `event.dispatcher.default.consumers` were run *in order*.  So, as long as the `discovery` consumer is listed _before_ the `submissionconfig` consumer, then shouldn't this Collection already be indexed?

Maybe I'm missing something though, or maybe these consumers are no longer run in order, or there's some sort of "race condition" here?

File: dspace/config/dspace.cfg [hunk 2]
```diff
@@ -780,11 +780,11 @@ event.dispatcher.default.class = org.dspace.event.BasicDispatcher
 # Add rdf here, if you are using dspace-rdf to export your repository content as RDF.
 # Add iiif here, if you are using dspace-iiif.
 # Add orcidqueue here, if the integration with ORCID is configured and wish to enable the synchronization queue functionality
-event.dispatcher.default.consumers = versioning, discovery, eperson
+event.dispatcher.default.consumers = versioning, discovery, eperson, submissionconfig
 
 # The noindex dispatcher will not create search or browse indexes (useful for batch item imports)
 event.dispatcher.noindex.class = org.dspace.event.BasicDispatcher
-event.dispatcher.noindex.consumers = eperson
+event.dispatcher.noindex.consumers = eperson, submissionconfig
```
Comment (tdonohue) [inline dspace/config/dspace.cfg:787]: Why are we added this new consumer to the `noindex` dispatcher?  If the `submissionconfig` needs to trigger reindexing of the Collection, then it shouldn't be listed here as it's writing to the index.

**Review by tdonohue — CHANGES_REQUESTED** [review]
@paulo-graca : Apologies for taking so long to get to this PR.  Overall, the code looks good (though I do have a few questions below).  But, I cannot get this to work exactly as planned.  Here's the behavior I'm seeing:

1. First, I made sure this PR is installed and I have the `submissionconfig` consumer configured and listed in the `event.dispatcher.default.consumers`.
2. Next, I created a brand new Collection.  I set it's Entity Type to "Person".
3. Then, I started a new "Person" entity submission.  I see this new Collection listed as one of my options to submit into.  So, I selected it.  (Up to here everything is working perfectly)
4. Unfortunately, though, in the newly created Submission, I see the Submission form for an **Item**.  I'm not shown the *Person* form.

I also tried the above process again for a new Collection of EntityType "OrgUnit". I had the same behavior.  While I *can* try to submit an OrgUnit into this new collection, the Submission form shown is always the default **Item** form.

Is this still working for you?  I'm running everything based on the latest `main`, but with the changes provided by this PR.

**Comment by paulo-graca** [comment]
> * First, I made sure this PR is installed and I have the `submissionconfig` consumer configured and listed in the `event.dispatcher.default.consumers`.
> * Next, I created a brand new Collection.  I set it's Entity Type to "Person".
> * Then, I started a new "Person" entity submission.  I see this new Collection listed as one of my options to submit into.  So, I selected it.  (Up to here everything is working perfectly)
> * Unfortunately, though, in the newly created Submission, I see the Submission form for an **Item**.  I'm not shown the _Person_ form.

I'm not sure if I forgot to mention in the description, @tdonohue did you also configured `item-submission.xml` and uncomment:
``` xml
        <name-map collection-entity-type="Publication" submission-name="Publication"/>
        <name-map collection-entity-type="Person" submission-name="Person"/>
        <name-map collection-entity-type="Project" submission-name="Project"/>
        <name-map collection-entity-type="OrgUnit" submission-name="OrgUnit"/>
...
```

**Comment by paulo-graca** [comment]
@tdonohue I've just tested again and it's working as expected:

image

File: dspace/config/dspace.cfg (hunk 2 above)
Comment (paulo-graca) [reply dspace/config/dspace.cfg]: Sorry, I've made several tests, including changing this configuration to understand why the collection wasn't available. I will undo this change that doesn't affect how it works.

File: dspace-api/src/main/java/org/dspace/submit/consumer/SubmissionConfigConsumer.java (hunk 1 above)
Comment (paulo-graca) [reply dspace-api/src/main/java/org/dspace/submit/consumer/SubmissionConfigConsumer.java:57]: I think it's related with the way IndexEventConsumer is implemented. The Solr `commit` will only occur in the `end` method, not in the `consume`:
https://github.com/paulo-graca/DSpace/blob/enhancement/itemsubmission-config-reload-consumer/dspace-api/src/main/java/org/dspace/discovery/IndexEventConsumer.java#L202-L234

**Comment by tdonohue** [comment]
Thanks @paulo-graca for verifying. I'm guessing I made a mistake in my initial configuration.  I'll retest this today and report back on my findings.

**Review by tdonohue — APPROVED** [review]
👍 Thanks @paulo-graca !  Retested today and verified it works.  I realized my mistake in earlier testing... I had a customized `event.dispatcher.default.consumers` listed in my `local.cfg` which did NOT include `submissionconfig` (so it overrode the default setup).  Once I corrected that, everything now works properly.

I've also verified that this is backwards compatible.  Everything works as before if the `submissionconfig` event consumer is missing or disabled.

**Comment by blancoj** [comment]
@paulo-graca thank you for giving me some guidance.  You are right that what I am trying to do is "having a form loaded according to a specific selection based on the user in session" ( as you said ).

So I have added a new metadata mapping field called collection_mappings and you can see it here:
https://github.com/mlibrary/DSpace/blob/issue-working/dspace/config/submission-forms.xml#L96

What I want to do is that DCInputsReader.java build a dropdown showing all the collections the user has access to deposit so that the user can map the item at ingest.  There is more code change to support that mapping, but for now what I want to do is to display the list of collections the user has access to deposit.  This code needs some cleaning up, but you can see here what I do in DCInputsReader.java here:

https://github.com/mlibrary/DSpace/blob/issue-working/dspace-api/src/main/java/org/dspace/app/util/DCInputsReader.java#L661

There is another field that I want to change based on the user logged in, but I know that if I can get this one to work, I can get the other one to work. 

What I am seeing is that this code in DCInputsReader.java runs when the user starts the deposit, but the UI only shows the form it initially created.  I cannot get the form to change.  It seems like the angular side grabs the 1st submission form created, caches it, and just uses that form.  What I want is for that form to be recreated every time the user starts a deposit.

Do you have any insight into this?  I have this working in 6.3, and I think at one time it was working in 7.4, but I can't get it to work in 7.6.  I think it has to be something on the angular side but have not had any luck finding it.

**Pull request merged**